### PR TITLE
Adding Translations for riot-web

### PR DESCRIPTION
**Pull request by MTRNord**
This is the one half of the translations other Part at: https://github.com/vector-im/riot-web/pull/3636

Both PR's depent on each other so it is required to merge them at the same time to get translations running

The translation is still in progress and I will cleanup before merging when it is finished.

## Current Status:

- [x] Language Setting
- [x] Important strings added to translation   (Not complete but will come after it. Also untranslated are english. Reason is we can't keep up with changing them in the same time as new strings are added)
- [x] Auto load Languages to settings
- [x] Language Default detection via browser
- [x] Language Fallback to en
- [x] Remove the not actually used react-translation-component dependency
- [x] fix the settings dropdown. (after selecting language it locks down to the loader.
- [x] Relaunch my public weblate platform for being able to add more translations easily (If wanted it will run "forever" to let anyone easily add new translations at any given time)
- [x] Add Docs for Translators
- [x] Add Docs for Developer
- [x] change counterpart.translate() to _t()
- [ ] Maybe add some Tests. (Propably not happening as it is not really possible to test. Counterpart doesn't really provide any way to check if it actually shows the correct value) Feel free to add it if you know how

Explanation: See docs dir at riot-web

### CODE REVIEW

**Comment by MTRNord** [comment]
Whoops Wrong base I will fix that

File: src/components/views/elements/LanguageDropdown.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
+/*
+Copyright 2017 Vector Creations Ltd
```
Comment (dbkr) [inline src/components/views/elements/LanguageDropdown.js]: You should assign the copyright of this file to yourself: you wrote it.

**Comment by dbkr** [comment]
This is looking great - some feedback before you go any further:

 * Be careful of the code style - we use 4 space indenting.
 * The TranslationsFix script should probably be done with webpack or something such that it automatically works with watch mode.
 * Be careful with your require syntax: we are switching to `import` in general. If an file hasn't been moved over to import yet, adding `require`s may be acceptable, but adding `require`s to a file with only imports is a no-no. If you do use `require`, it should be with `const`, not `var`.
 * We will have to pull in the translations files asynchronously when the language is set: we can't include all the translations in the bundle.

Other than that, this is looking very promising!

File: src/components/views/elements/LanguageDropdown.js (hunk 1 above)
Comment (MTRNord) [inline src/components/views/elements/LanguageDropdown.js]: @dbkr Nah I copied it :)  Also it will be replaced within the next commits

**Comment by MTRNord** [comment]
@dbkr i will do a code cleanup commit after everythink works.

**Comment by MTRNord** [comment]
@dbkr react-sdk uses babel for compiling so there is no real way to copy files. I will see if I can a watching script that retriggers `TranslationsFix.js` as I don't see a better way to do this yet. on riot-web side it works already with webpack out of the box.

"We will have to pull in the translations files asynchronously when the language is set: we can't include all the translations in the bundle."  --> What do you have in mind about this? Do you mean only do the `registerTranslations` for the current language and when changing  the language?

**Comment by dtygel** [comment]
Hi all. We've prepared more files to be able to be translated. Apart of that, we've added the new strings to English and Portuguese (Brazilian):

https://github.com/MTRNord/matrix-react-sdk/pull/10

**Comment by MTRNord** [comment]
Edit: added some explanation what logic is added.

**Comment by MTRNord** [comment]
To discuss about the weblate instance is if a webhook should be added. It would do than auto update weblate as soon a new commit happens. Doc would be here: https://docs.weblate.org/en/latest/admin/continuous.html#automatically-receiving-changes-from-github  At a later point I can provide the needed payload URL

File: src/components/views/elements/LanguageDropdown.js (hunk 1 above)
Comment (MTRNord) [inline src/components/views/elements/LanguageDropdown.js]: @dbkr done :)

**Comment by MTRNord** [comment]
Please ping me as soon as you mere it so that I can update the weblate config

File: src/components/views/login/PasswordLogin.js [hunk 2]
```diff
@@ -199,24 +200,24 @@ class PasswordLogin extends React.Component {
...
-                    <label className="mx_Login_type_label">I want to sign in with my</label>
+                    <label className="mx_Login_type_label">{ counterpart.translate('I want to sign in with') }</label>
                     <Dropdown
                         className="mx_Login_type_dropdown"
                         value={this.state.loginType}
                         onOptionChange={this.onLoginTypeChange}>
-                            <span key={PasswordLogin.LOGIN_FIELD_MXID}>Matrix ID</span>
-                            <span key={PasswordLogin.LOGIN_FIELD_EMAIL}>Email Address</span>
-                            <span key={PasswordLogin.LOGIN_FIELD_PHONE}>Phone</span>
+                            <span key={PasswordLogin.LOGIN_FIELD_MXID}>{ counterpart.translate('my Matrix ID') }</span>
```
Comment (krombel) [inline src/components/views/login/PasswordLogin.js]: I would move "my" to the previous string (where it was before) as it fits to the other options as well. And it avoids confusion because "my" is just there in combination with Matrix ID

File: src/components/views/rooms/MessageComposer.js [hunk 3]
```diff
@@ -298,7 +298,7 @@ export default class MessageComposer extends React.Component {
             );
 
             const placeholderText = roomIsEncrypted ?
-                "Send an encrypted message…" : "Send a message (unencrypted)…";
+                counterpart.translate("Send an encrypted message…") : counterpart.translate("Send a message (unencrypted)…");
```
Comment (krombel) [inline src/components/views/rooms/MessageComposer.js]: Suggest: do not translate "…"

File: src/components/views/rooms/RoomPreviewBar.js [hunk 4]
```diff
@@ -131,10 +133,10 @@ module.exports = React.createClass({
             joinBlock = (
                 <div>
                     <div className="mx_RoomPreviewBar_invite_text">
-                        You have been invited to join this room by <b>{ this.props.inviterName }</b>
+                        { counterpart.translate("You have been invited to join this room by") } <b>{ this.props.inviterName }</b>
```
Comment (krombel) [inline src/components/views/rooms/RoomPreviewBar.js]: Suggestion: Add inviterName as paramater for the translation

File: src/i18n/strings/ru.json [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+    "changed their display name from": "changed their display name from",
+    "changed their profile picture": "changed their profile picture",
+    "changed the power level of": "changed the power level of",
+    "changed the room name to": "changed the room name to",
+    "changed the topic to": "changed the topic to",
+    "Changes to who can read history will only apply to future messages in this room": "",
+    "Changes your display nickname": "",
+    "Claimed Ed25519 fingerprint key": "",
+    "Clear Cache and Reload": "",
+    "Clear Cache": "",
```
Comment (krombel) [inline src/i18n/strings/ru.json]: Can you remove this empty translations? Not that the string is missing completely later...

**Comment by MTRNord** [comment]
@krombel about the ru translations i saw that too i think I need to add a Note about it in the docs... I will fix that

**Comment by MTRNord** [comment]
About the failing test: the Test requires the PR over at riot-web to work. Otherise it will fail loading lang files because the webser doesn't serve required json files

File: src/TextForEvent.js [hunk 6]
```diff
@@ -25,95 +25,91 @@ function textForMemberEvent(ev) {
...
+                return counterpart.translate("%(targetName)s left the room", {targetName: targetName}) + ".";
             }
     }
 }
 
 function textForTopicEvent(ev) {
     var senderDisplayName = ev.sender && ev.sender.name ? ev.sender.name : ev.getSender();
-
-    return senderDisplayName + ' changed the topic to "' + ev.getContent().topic + '"';
+    return counterpart.translate("%(senderDisplayName)s changed the topic to %(topic)s", {senderDisplayName: senderDisplayName, topic: ev.getContent().topic});
```
Comment (MTRNord) [inline src/TextForEvent.js]: Done

File: src/autocomplete/CommandProvider.js [hunk 7]
```diff
@@ -78,7 +79,7 @@ export default class CommandProvider extends AutocompleteProvider {
     }
 
     getName() {
-        return '*️⃣ Commands';
+        return counterpart.translate('*️⃣ Commands');
```
Comment (MTRNord) [inline src/autocomplete/CommandProvider.js]: done

File: src/autocomplete/EmojiProvider.js [hunk 8]
```diff
@@ -39,7 +40,7 @@ export default class EmojiProvider extends AutocompleteProvider {
     }
 
     getName() {
-        return '😃 Emoji';
+        return counterpart.translate('😃 Emoji');
```
Comment (MTRNord) [inline src/autocomplete/EmojiProvider.js]: done

File: src/autocomplete/RoomProvider.js [hunk 9]
```diff
@@ -50,7 +51,7 @@ export default class RoomProvider extends AutocompleteProvider {
     }
 
     getName() {
-        return '💬 Rooms';
+        return counterpart.translate('💬 Rooms');
```
Comment (MTRNord) [inline src/autocomplete/RoomProvider.js]: done

File: src/autocomplete/UserProvider.js [hunk 10]
```diff
@@ -51,7 +52,7 @@ export default class UserProvider extends AutocompleteProvider {
     }
 
     getName() {
-        return '👥 Users';
+        return counterpart.translate('👥 Users');
```
Comment (MTRNord) [inline src/autocomplete/UserProvider.js]: done

File: src/components/structures/login/Login.js [hunk 11]
```diff
@@ -223,14 +224,14 @@ module.exports = React.createClass({
                  !this.state.enteredHomeserverUrl.startsWith("http")))
             {
                 errorText = <span>
-                    Can't connect to homeserver via HTTP when an HTTPS URL is in your browser bar.
-                    Either use HTTPS or <a href='https://www.google.com/search?&q=enable%20unsafe%20scripts'>enable unsafe scripts</a>
+                    { counterpart.translate("Can't connect to homeserver via HTTP when an HTTPS URL is in your browser bar")}.
+                    { counterpart.translate('Either use HTTPS or ')}<a href='https://www.google.com/search?&q=enable%20unsafe%20scripts'>{ counterpart.translate('enable unsafe scripts')}</a>
```
Comment (MTRNord) [inline src/components/structures/login/Login.js]: @krombel hm not sure about it as google should autodetect the browser language. and riot does the requests via the client so that should work for itself.

File: src/components/structures/login/Login.js [hunk 12]
```diff
@@ -223,14 +224,14 @@ module.exports = React.createClass({
...
                 </span>;
             }
             else {
                 errorText = <span>
-                    Can't connect to homeserver - please check your connectivity and ensure
-                    your <a href={ this.state.enteredHomeserverUrl }>homeserver's SSL certificate</a> is trusted.
+                    { counterpart.translate("Can't connect to homeserver - please check your connectivity and ensure your")} 
+                    <a href={ this.state.enteredHomeserverUrl }>{ counterpart.translate("homeserver's SSL certificate")}</a> { counterpart.translate('is trusted') }.
```
Comment (MTRNord) [inline src/components/structures/login/Login.js]: done

File: src/components/views/login/PasswordLogin.js (hunk 2 above)
Comment (MTRNord) [inline src/components/views/login/PasswordLogin.js]: @krombel but I guess the my is there context related or so

File: src/components/views/rooms/MessageComposer.js (hunk 3 above)
Comment (MTRNord) [inline src/components/views/rooms/MessageComposer.js]: done

File: src/components/views/rooms/RoomPreviewBar.js (hunk 4 above)
Comment (MTRNord) [inline src/components/views/rooms/RoomPreviewBar.js]: done

File: src/i18n/strings/ru.json (hunk 5 above)
Comment (MTRNord) [inline src/i18n/strings/ru.json]: done

File: package.json [hunk 13]
```diff
@@ -50,6 +50,7 @@
     "browser-request": "^0.3.3",
     "classnames": "^2.1.2",
     "commonmark": "^0.27.0",
+    "counterpart": "Nordgedanken/counterpart#develop",
```
Comment (ara4n) [inline package.json]: please can we depend on a specific revision rather than just #develop?

File: src/DateUtils.js [hunk 14]
```diff
@@ -15,9 +15,36 @@ limitations under the License.
 */
 
 'use strict';
+import _t from 'counterpart';
 
-var days = ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"];
-var months = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];
+function getDaysArray() {
+	var days = [];
+	days.push(_t('Sun'));
```
Comment (ara4n) [inline src/DateUtils.js]: i'm still very suspicious that the array literal doesn't internationalise correctly.  Perhaps it's an autohoisting thing - does it work if you do `var days;` and then `days = [ _t('Mon'), ... ]`?

Comment (ara4n) [inline src/DateUtils.js]: failing that, this definitely needs a comment to explain why the weirdness is happening :)

File: src/DateUtils.js [hunk 15]
```diff
@@ -27,28 +54,39 @@ module.exports = {
     formatDate: function(date) {
         // date.toLocaleTimeString is completely system dependent.
         // just go 24h for now
+        function pad(n) {
+            return (n < 10 ? '0' : '') + n;
+        }
```
Comment (ara4n) [inline src/DateUtils.js]: this is a bad merge - we shouldn't be defining pad on line 49 as well as 57

File: src/DateUtils.js [hunk 16]
```diff
@@ -27,28 +54,39 @@ module.exports = {
...
+        const days = getDaysArray();
+        const months = getMonthsArray();
 
+		// TODO: use standard date localize function provided in counterpart
+		var hoursAndMinutes = pad(date.getHours()) + ':' + pad(date.getMinutes());
```
Comment (ara4n) [inline src/DateUtils.js]: indentation seems worryingly broken here - are these tabs?! if so, please please please go through and turn the tabs into 4-spaces please in accordance to the codestyle :)

File: src/Lifecycle.js [hunk 17]
```diff
@@ -229,14 +230,12 @@ function _handleRestoreFailure(e) {
 
     let msg = e.message;
     if (msg == "OLM.BAD_LEGACY_ACCOUNT_PICKLE") {
-        msg = "You need to log back in to generate end-to-end encryption keys "
-            + "for this device and submit the public key to your homeserver. "
-            + "This is a once off; sorry for the inconvenience.";
+        msg = _t('You need to log back in to generate end-to-end encryption keys for this device and submit the public key to your homeserver. This is a once off; sorry for the inconvenience') + '.';
```
Comment (ara4n) [inline src/Lifecycle.js]: it'd be nice to keep long messages like this split over multiple lines if possible for legibility (and per the codestyle)

File: src/components/structures/RoomView.js [hunk 18]
```diff
@@ -1,4 +1,4 @@
-/*
+_t/*
```
Comment (ara4n) [inline src/components/structures/RoomView.js]: erm, what is this _t doing?

File: src/components/structures/RoomView.js [hunk 19]
```diff
@@ -813,7 +815,7 @@ module.exports = React.createClass({
             if (this.state.room) {
                 const me = this.state.room.getMember(MatrixClientPeg.get().credentials.userId);
                 if (me && me.membership == 'invite') {
-                    if (me.events.member.getContent().is_direct) {
+                    if (me.events.member.getContent().is_direct && 0) { // eliminate direct chat sabotage
```
Comment (ara4n) [inline src/components/structures/RoomView.js]: hm, this feels unrelated; what's it doing here?

File: src/components/structures/UserSettings.js [hunk 20]
```diff
@@ -191,11 +197,21 @@ module.exports = React.createClass({
 
         const syncedSettings = UserSettingsStore.getSyncedSettings();
         if (!syncedSettings.theme) {
-            syncedSettings.theme = 'light';
+            syncedSettings.theme = 'cadcampo';
```
Comment (ara4n) [inline src/components/structures/UserSettings.js]: pretty sure that shouldn't be there ;)

File: src/components/structures/UserSettings.js [hunk 21]
```diff
@@ -191,11 +197,21 @@ module.exports = React.createClass({
...
         }
         this._syncedSettings = syncedSettings;
 
         this._localSettings = UserSettingsStore.getLocalSettings();
+        if (!this._localSettings.hasOwnProperty('language')) {
```
Comment (ara4n) [inline src/components/structures/UserSettings.js]: please fix the indentation to the code style (4 space indenting, and `} else {`

File: src/components/structures/login/ForgotPassword.js [hunk 22]
```diff
@@ -78,30 +79,27 @@ module.exports = React.createClass({
...
             Modal.createDialog(QuestionDialog, {
-                title: "Warning",
+                title: _t('Warning'),
                 description:
                     <div>
-                        Resetting password will currently reset any end-to-end encryption keys on all devices,
-                        making encrypted chat history unreadable, unless you first export your room keys
-                        and re-import them afterwards.
-                        In future this <a href="https://github.com/vector-im/riot-web/issues/2671">will be improved</a>.
+                        { _t('Resetting password will currently reset any end-to-end encryption keys on all devices, making encrypted chat history unreadable, unless you first export your room keys and re-import them afterwards. In future this <a href="https://github.com/vector-im/riot-web/issues/2671">will be improved</a>') }.
```
Comment (ara4n) [inline src/components/structures/login/ForgotPassword.js]: please wrap this again :)

File: src/components/structures/login/ForgotPassword.js [hunk 23]
```diff
@@ -150,7 +148,8 @@ module.exports = React.createClass({
         var ErrorDialog = sdk.getComponent("dialogs.ErrorDialog");
         Modal.createDialog(ErrorDialog, {
             title: title,
-            description: body
+            description: body,
+            button: _t("OK"),
```
Comment (ara4n) [inline src/components/structures/login/ForgotPassword.js]: i think the "OK" i18n should be done on BaseDialog itself, to avoid having to repeat it everywhere?

File: src/components/views/dialogs/ChatInviteDialog.js [hunk 24]
```diff
@@ -48,11 +49,7 @@ module.exports = React.createClass({
 
     getDefaultProps: function() {
         return {
```
Comment (ara4n) [inline src/components/views/dialogs/ChatInviteDialog.js]: hm, where did these props go?

Comment (ara4n) [inline src/components/views/dialogs/ChatInviteDialog.js]: oh, i see - presumably they're now passed in as the defaults can't be i18nized

File: src/components/views/elements/MemberEventListSummary.js [hunk 25]
```diff
@@ -206,27 +207,27 @@ module.exports = React.createClass({
...
+            "banned": beConjugated + ' ' + _t('was banned'),
+            "unbanned": beConjugated + ' ' + _t('was unbanned'),
+            "kicked": beConjugated + ' ' + _t('was kicked'),
+            "changed_name": _t('changed name'),
+            "changed_avatar": _t('changed avatar'),
         };
 
         if (Object.keys(map).includes(t)) {
-            res = map[t] + (repeats > 1 ? " " + repeats + " times" : "" );
+            res = map[t] + (repeats > 1 ? ' ' + repeats + ' ' + _t('times') : '' );
```
Comment (ara4n) [inline src/components/views/elements/MemberEventListSummary.js]: i guess this all assumes "noun, verb, adverb" word ordering as per English... i wonder how many languages this will cause a problem for...

File: src/components/views/rooms/MessageComposer.js [hunk 26]
```diff
@@ -116,15 +116,16 @@ export default class MessageComposer extends React.Component {
...
             description: (
                 <div>
-                    <p>Are you sure you want upload the following files?</p>
+                    <p>{ _t('Are you sure you want upload the following files?') }</p>
                     <ul style={{listStyle: 'none', textAlign: 'left'}}>
                         {fileList}
                     </ul>
                 </div>
             ),
+            button: "OK",
```
Comment (ara4n) [inline src/components/views/rooms/MessageComposer.js]: missing _t()? and should be provided by BaseDialog surely?

File: src/languageHandler.js [hunk 27]
```diff
@@ -0,0 +1,166 @@
+/*
+Copyright 2015, 2016 OpenMarket Ltd
```
Comment (ara4n) [inline src/languageHandler.js]: should use the right (C).  also needs major indentation fixing :)

**Comment by ara4n** [comment]
right - i've gone and read the whole thing; this is a *spectacular* amount of work; congratulations on getting pretty much the whole app covered!!

There are some relatively minor bits of feedback above.   Separately, I haven't properly dug into LanguageDropdown and languageHandler yet; @dbkr have you taken a look?

File: src/components/structures/login/ForgotPassword.js (hunk 23 above)
Comment (dtygel) [inline src/components/structures/login/ForgotPassword.js]: The problem is that OK is not an universal word, and strings in getDefaultProps won't get translated, since they are set before language is loaded. So instead of stopping everything to load to wait languages to load, we decided to suggest developers that getDefaultProps don't have translateable strings.

File: src/components/views/dialogs/ChatInviteDialog.js (hunk 24 above)
Comment (dtygel) [inline src/components/views/dialogs/ChatInviteDialog.js]: Exactly. This explains the 'OK' issue.

**Comment by dtygel** [comment]
@ara4n Thanks for the quick and qualified and precise review! I've done a PR ( https://github.com/MTRNord/matrix-react-sdk/pull/89 ) to @MTRNord translations branch, which fixes all the issues pointed by you (at least I hope so). So, after @MTRNord reviews my PR to his branch, they will be here for you to double check.

File: src/languageHandler.js (hunk 27 above)
Comment (dtygel) [inline src/languageHandler.js]: Fixed. @MTRNord , should we share (c) for the languageHandler? Or do do you prefer to have it only in your name? It's up to you! You will see that I've put "NordGedanken and Cooperative EITA", but it can be whatever you want, since you really made it!

File: src/components/views/rooms/MessageComposer.js (hunk 26 above)
Comment (dtygel) [inline src/components/views/rooms/MessageComposer.js]: Solved, but unfortunately, the OK must be in the caller, since in the basedialog the language wasn't yet loaded...

File: src/components/views/elements/MemberEventListSummary.js (hunk 25 above)
Comment (dtygel) [inline src/components/views/elements/MemberEventListSummary.js]: You are right. I solved it with a biiiig code, to respect word orderings of any language. The only thing I left intact was the list of user names before the transition phrase.

File: src/components/structures/login/ForgotPassword.js (hunk 22 above)
Comment (dtygel) [inline src/components/structures/login/ForgotPassword.js]: I'm not sure if wrapping is a good thing. This will make it more difficult for a script to auto-add strings to en_EN.json. But I've wrapped it anyway, and we can discuss this later...

File: src/components/structures/UserSettings.js (hunk 21 above)
Comment (dtygel) [inline src/components/structures/UserSettings.js]: bad 2 spacings... Done.

File: src/components/structures/UserSettings.js (hunk 20 above)
Comment (dtygel) [inline src/components/structures/UserSettings.js]: Oh no, sabotage again! Fixed.

File: src/components/structures/RoomView.js (hunk 19 above)
Comment (dtygel) [inline src/components/structures/RoomView.js]: Oh no, sabotage! Fixed.

File: src/components/structures/RoomView.js (hunk 18 above)
Comment (dtygel) [inline src/components/structures/RoomView.js]: Ooops, damn ctrl-c. Fixed.

File: src/Lifecycle.js (hunk 17 above)
Comment (dtygel) [inline src/Lifecycle.js]: done.

File: src/DateUtils.js (hunk 16 above)
Comment (dtygel) [inline src/DateUtils.js]: fixed as well.

File: src/DateUtils.js (hunk 15 above)
Comment (dtygel) [inline src/DateUtils.js]: yep, sorry for that... fixed.

**Comment by MTRNord** [comment]
@dtygel done @ara4n packageg json updated to use counterpart-riot wich I now did publish

**Comment by MTRNord** [comment]
I now need to update the imports

**Comment by dbkr** [comment]
Merged manually in https://github.com/matrix-org/matrix-react-sdk/pull/934
